domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,106 @@
+internal
+class QuestionAssessmentCalculation @Inject constructor(
+  @ViewHintPenalty private val viewHintPenalty: Double,
+  @WrongAnswerPenalty private val wrongAnswerPenalty: Double,
+  @MaxScorePerQuestion private val maxScorePerQuestion: Double
+) {
+  private lateinit var skillIdList: List<String>
+  private lateinit var questionSessionMetrics: List<QuestionSessionMetrics>
+  private lateinit var totalScore: MutableFractionGrade
+  private lateinit var scorePerSkillMapping: MutableMap<String, MutableFractionGrade>
+
+  /** Initialize member fields. */
+  internal fun initialize(
+    skillIdList: List<String>,
+    questionSessionMetrics: List<QuestionSessionMetrics>
+  ) {
+    this.skillIdList = skillIdList
+    this.questionSessionMetrics = questionSessionMetrics
+    this.totalScore =
+      MutableFractionGrade(0.toBigDecimal(), questionSessionMetrics.size.toBigDecimal())
+    createScorePerSkillMapping()
+  }
+
+  /** Initialize the scorePerSkillMapping member field. */
+  private fun createScorePerSkillMapping() {
+    scorePerSkillMapping = mutableMapOf()
+    for (skillId in skillIdList) {
+      scorePerSkillMapping[skillId] = MutableFractionGrade(0.toBigDecimal(), 0.toBigDecimal())
+    }
+  }
+
+  /** Calculate the user's overall score and score per skill for this practice session. */
+  private fun calculateScores() {
+    for (questionMetric in questionSessionMetrics) {
+      val totalHintsPenalty =
+        questionMetric.numberOfHintsUsed.toBigDecimal() * viewHintPenalty.toBigDecimal()
+      val totalWrongAnswerPenalty =
+        (questionMetric.numberOfAnswersSubmitted - 1)
+          .toBigDecimal() * wrongAnswerPenalty.toBigDecimal()
+      var questionScore = 0.toBigDecimal()
+      if (!questionMetric.didViewSolution) {
+        questionScore =
+          maxOf(
+            maxScorePerQuestion.toBigDecimal() - totalHintsPenalty - totalWrongAnswerPenalty,
+            questionScore
+          )
+      }
+      this.totalScore.numerator += questionScore
+      for (linkedSkillId in questionMetric.question.linkedSkillIdsList) {
+        if (!scorePerSkillMapping.containsKey(linkedSkillId)) continue
+        scorePerSkillMapping[linkedSkillId]!!.numerator += questionScore
+        scorePerSkillMapping[linkedSkillId]!!.denominator += 1.toBigDecimal()
+      }
+    }
+  }
+
+  /** Compute the overall score as well as the score and mastery per skill. */
+  internal fun computeAll(): UserAssessmentPerformance {
+    calculateScores()
+    // TODO(#3067): create mastery calculations method
+
+    // Set up the return values
+    var finalScore = FractionGrade.newBuilder().apply {
+      numerator = totalScore.numerator.toDouble()
+      denominator = totalScore.denominator.toDouble()
+    }.build()
+
+    var finalScorePerSkillMapping = mutableMapOf<String, FractionGrade>()
+    for (score in scorePerSkillMapping) {
+      if (score.value.denominator != 0.toBigDecimal()) { // Exclude entries with denominator of 0
```

Comment:
@vinitamurthi We're removing entries with 0 numerator/denominator in the computeAll() method here